[JIT] Optionally validate nvfuser outputs after execution

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75116
* #71299
* #73322
* __->__ #74361

This adds an optional validation after executing an NVFuser node, which checks that the output is the same as the unfused implementation. Then the outputs and the graph are reported via a callback.

```python
import torch

def callback(x, y, graph):
    for i in range(len(x)-amt, len(x)):
        print(x[i])
        print(y[i])
    print(graph)


with torch.jit.fuser("fuser2"):
    torch._C._jit_nvfuser_set_comparison_callback(True, callback)

    @torch.jit.script
    def g(x, y):
        z = torch.add(x, y)
        return torch.sin(z)

    def f(x, y, a):
        z = torch.add(x, y)
        return g(torch.relu(z), a)

    f_s = torch.jit.script(f)
    x = torch.rand((10, 10), dtype=torch.half).cuda()
    y = torch.rand((10, 10), dtype=torch.half).cuda()
    a = torch.rand((10, 10), dtype=torch.half).cuda()
    f_s(x, y, a)
    f_s(x, y, a)
    f_s(x, y, a)
```

Differential Revision: [D34975310](https://our.internmc.facebook.com/intern/diff/D34975310)

davidberard98 committed:
[WIP][JIT] Optionally validate nvfuser outputs after execution

This adds an optional validation after executing an NVFuser node, which checks that the output is the same as the unfused implementation.o

```python
import torch

def callback(amt, x, y):
    for i in range(len(x)-amt, len(x)):
        print(x[i])
        print(y[i])

with torch.jit.fuser("fuser2"):
    torch._C._jit_nvfuser_set_comparison_callback(callback)

    @torch.jit.script
    def g(x, y):
        z = torch.add(x, y)
        return torch.sin(z)

    def f(x, y, a):
        z = torch.add(x, y)
        return g(torch.relu(z), a)

    f_s = torch.jit.script(f)
    x = torch.rand((10, 10), dtype=torch.half).cuda()
    y = torch.rand((10, 10), dtype=torch.half).cuda()
    a = torch.rand((10, 10), dtype=torch.half).cuda()
    f_s(x, y, a)
    f_s(x, y, a)
    f_s(x, y, a)
```

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] Optionally validate nvfuser outputs after execution"


This adds an optional validation after executing an NVFuser node, which checks that the output is the same as the unfused implementation.o

```python
import torch

def callback(amt, x, y):
    for i in range(len(x)-amt, len(x)):
        print(x[i])
        print(y[i])

with torch.jit.fuser("fuser2"):
    torch._C._jit_nvfuser_set_comparison_callback(callback)

    torch.jit.script
    def g(x, y):
        z = torch.add(x, y)
        return torch.sin(z)

    def f(x, y, a):
        z = torch.add(x, y)
        return g(torch.relu(z), a)

    f_s = torch.jit.script(f)
    x = torch.rand((10, 10), dtype=torch.half).cuda()
    y = torch.rand((10, 10), dtype=torch.half).cuda()
    a = torch.rand((10, 10), dtype=torch.half).cuda()
    f_s(x, y, a)
    f_s(x, y, a)
    f_s(x, y, a)
```

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34975310).

davidberard98 committed:
Update on "[WIP][JIT] Optionally validate nvfuser outputs after execution"


This adds an optional validation after executing an NVFuser node, which checks that the output is the same as the unfused implementation.o

```python
import torch

def callback(amt, x, y):
    for i in range(len(x)-amt, len(x)):
        print(x[i])
        print(y[i])

with torch.jit.fuser("fuser2"):
    torch._C._jit_nvfuser_set_comparison_callback(callback)

    torch.jit.script
    def g(x, y):
        z = torch.add(x, y)
        return torch.sin(z)

    def f(x, y, a):
        z = torch.add(x, y)
        return g(torch.relu(z), a)

    f_s = torch.jit.script(f)
    x = torch.rand((10, 10), dtype=torch.half).cuda()
    y = torch.rand((10, 10), dtype=torch.half).cuda()
    a = torch.rand((10, 10), dtype=torch.half).cuda()
    f_s(x, y, a)
    f_s(x, y, a)
    f_s(x, y, a)
```

Differential Revision: [D34975310](https://our.internmc.facebook.com/intern/diff/D34975310)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Optionally validate nvfuser outputs after execution"


This adds an optional validation after executing an NVFuser node, which checks that the output is the same as the unfused implementation. Then the outputs and the graph are reported via a callback.

```python
import torch

def callback(x, y, graph):
    for i in range(len(x)-amt, len(x)):
        print(x[i])
        print(y[i])
    print(graph)


with torch.jit.fuser("fuser2"):
    torch._C._jit_nvfuser_set_comparison_callback(True, callback)

    torch.jit.script
    def g(x, y):
        z = torch.add(x, y)
        return torch.sin(z)

    def f(x, y, a):
        z = torch.add(x, y)
        return g(torch.relu(z), a)

    f_s = torch.jit.script(f)
    x = torch.rand((10, 10), dtype=torch.half).cuda()
    y = torch.rand((10, 10), dtype=torch.half).cuda()
    a = torch.rand((10, 10), dtype=torch.half).cuda()
    f_s(x, y, a)
    f_s(x, y, a)
    f_s(x, y, a)
```

Differential Revision: [D34975310](https://our.internmc.facebook.com/intern/diff/D34975310)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34975310).

davidberard98 committed:
Update on "[JIT] Optionally validate nvfuser outputs after execution"


This adds an optional validation after executing an NVFuser node, which checks that the output is the same as the unfused implementation. Then the outputs and the graph are reported via a callback.

```python
import torch

def callback(x, y, graph):
    for i in range(len(x)-amt, len(x)):
        print(x[i])
        print(y[i])
    print(graph)


with torch.jit.fuser("fuser2"):
    torch._C._jit_nvfuser_set_comparison_callback(True, callback)

    torch.jit.script
    def g(x, y):
        z = torch.add(x, y)
        return torch.sin(z)

    def f(x, y, a):
        z = torch.add(x, y)
        return g(torch.relu(z), a)

    f_s = torch.jit.script(f)
    x = torch.rand((10, 10), dtype=torch.half).cuda()
    y = torch.rand((10, 10), dtype=torch.half).cuda()
    a = torch.rand((10, 10), dtype=torch.half).cuda()
    f_s(x, y, a)
    f_s(x, y, a)
    f_s(x, y, a)
```

Differential Revision: [D34975310](https://our.internmc.facebook.com/intern/diff/D34975310)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34975310).

eellison commented on torch/csrc/jit/codegen/cuda/manager.cpp:
```diff
@@ -209,14 +210,32 @@ void runCudaFusionGroup(const Node* fusion_node, Stack& stack) {
     auto copied_graph = fusion_node->g(attr::Subgraph)->copy();
     EraseShapeInformation(copied_graph);
     enableAliasCopyNodes(copied_graph, copied_graph->block());
     InterpreterState{Code(copied_graph, "fallback_cuda_fuser")}.run(stack);
```
We should really cache and save the `InterpreterState` code here. This fallback actually does get run in real networks, and this whole processes happens on each invocation

eellison replied:
That probably shouldn't be in this pr - but something to do soon

eellison commented on torch/csrc/jit/passes/cuda_graph_fuser.h:
```diff
@@ -29,5 +31,15 @@ struct TORCH_API RegisterCudaFuseGraph
+struct CudaFuserComparisonCallback {
+  using callback_type =
+      std::function<void(const Stack&, const Stack&, const std::string&)>;
+  bool run_fallback;
```
When would this ever be `false` ?

davidberard98 replied:
@eellison if you just want to collect the graph (e.g. for logging fusion groups), without running the fallback

eellison reviewed (approved):
Sweet!! LGTM - just a couple small comments. CC @jjsjann123 @kevinstephano may want to review as well

davidberard98 committed:
Update on "[JIT] Optionally validate nvfuser outputs after execution"


This adds an optional validation after executing an NVFuser node, which checks that the output is the same as the unfused implementation. Then the outputs and the graph are reported via a callback.

```python
import torch

def callback(x, y, graph):
    for i in range(len(x)-amt, len(x)):
        print(x[i])
        print(y[i])
    print(graph)


with torch.jit.fuser("fuser2"):
    torch._C._jit_nvfuser_set_comparison_callback(True, callback)

    torch.jit.script
    def g(x, y):
        z = torch.add(x, y)
        return torch.sin(z)

    def f(x, y, a):
        z = torch.add(x, y)
        return g(torch.relu(z), a)

    f_s = torch.jit.script(f)
    x = torch.rand((10, 10), dtype=torch.half).cuda()
    y = torch.rand((10, 10), dtype=torch.half).cuda()
    a = torch.rand((10, 10), dtype=torch.half).cuda()
    f_s(x, y, a)
    f_s(x, y, a)
    f_s(x, y, a)
```

Differential Revision: [D34975310](https://our.internmc.facebook.com/intern/diff/D34975310)

[ghstack-poisoned]

jjsjann123 commented on torch/csrc/jit/codegen/cuda/manager.cpp:
```diff
@@ -209,14 +210,32 @@ void runCudaFusionGroup(const Node* fusion_node, Stack& stack) {
+    TORCH_INTERNAL_ASSERT(stack.size() >= inputs_size);
+    stack_copy = Stack();
+    stack_copy->insert(
+        stack_copy->end(), stack.begin(), stack.end() - inputs_size);
```
Should we assert on `!stack_copy.empty()`? Since we later (L297) only use to determine whether we re-run fallback.

davidberard98 replied:
stack_copy is a c10::optional<> so conversion to bool will return true as long as it has a value - I can change the later if statement to `if (stack_copy.has_value())` for clarity

jjsjann123 replied:
sorry I missed the `optional` part... code looks fine as-is.

jjsjann123 reviewed (approved):
LGTM

davidberard98 committed:
Update on "[JIT] Optionally validate nvfuser outputs after execution"


This adds an optional validation after executing an NVFuser node, which checks that the output is the same as the unfused implementation. Then the outputs and the graph are reported via a callback.

```python
import torch

def callback(x, y, graph):
    for i in range(len(x)-amt, len(x)):
        print(x[i])
        print(y[i])
    print(graph)


with torch.jit.fuser("fuser2"):
    torch._C._jit_nvfuser_set_comparison_callback(True, callback)

    torch.jit.script
    def g(x, y):
        z = torch.add(x, y)
        return torch.sin(z)

    def f(x, y, a):
        z = torch.add(x, y)
        return g(torch.relu(z), a)

    f_s = torch.jit.script(f)
    x = torch.rand((10, 10), dtype=torch.half).cuda()
    y = torch.rand((10, 10), dtype=torch.half).cuda()
    a = torch.rand((10, 10), dtype=torch.half).cuda()
    f_s(x, y, a)
    f_s(x, y, a)
    f_s(x, y, a)
```

Differential Revision: [D34975310](https://our.internmc.facebook.com/intern/diff/D34975310)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Optionally validate nvfuser outputs after execution"


This adds an optional validation after executing an NVFuser node, which checks that the output is the same as the unfused implementation. Then the outputs and the graph are reported via a callback.

```python
import torch

def callback(x, y, graph):
    for i in range(len(x)-amt, len(x)):
        print(x[i])
        print(y[i])
    print(graph)


with torch.jit.fuser("fuser2"):
    torch._C._jit_nvfuser_set_comparison_callback(True, callback)

    torch.jit.script
    def g(x, y):
        z = torch.add(x, y)
        return torch.sin(z)

    def f(x, y, a):
        z = torch.add(x, y)
        return g(torch.relu(z), a)

    f_s = torch.jit.script(f)
    x = torch.rand((10, 10), dtype=torch.half).cuda()
    y = torch.rand((10, 10), dtype=torch.half).cuda()
    a = torch.rand((10, 10), dtype=torch.half).cuda()
    f_s(x, y, a)
    f_s(x, y, a)
    f_s(x, y, a)
```

Differential Revision: [D34975310](https://our.internmc.facebook.com/intern/diff/D34975310)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34975310).
